src/ansiblelint/transformer.py

```diff
@@ -0,0 +1,227 @@
+            yaml.map_indent=2
+            yaml.sequence_indent=4
+            yaml.sequence_dash_offset=2
+            yaml.Emitter = FormattedEmitter  # custom Emitter prevents root-level indents
+
+        .. code:: yaml
+
+            ---
+            - name: playbook
+              tasks:
+                - name: task
+        """
+        if self._yaml is not None:
+            return self._yaml
+
+        # ruamel.yaml rt=round trip (preserves comments while allowing for modification)
+        yaml = YAML(typ="rt")
+
+        # NB: ruamel.yaml does not have typehints, so mypy complains about everything here.
+
+        # configure yaml dump formatting
+        yaml.explicit_start = self.yaml_explicit_start  # type: ignore[assignment]
+        yaml.explicit_end = self.yaml_explicit_end  # type: ignore[assignment]
+        yaml.width = self.yaml_width  # type: ignore[assignment]
+
+        yaml.default_flow_style = False
+        yaml.compact_seq_seq = (  # dash after dash
+            True  # type: ignore[assignment]
+        )
+        yaml.compact_seq_map = (  # key after dash
+            True  # type: ignore[assignment]
+        )
+        # Do not use yaml.indent() as it obscures the purpose of these vars:
+        yaml.map_indent = 2  # type: ignore[assignment]
+        yaml.sequence_indent = 4 if self.yaml_indent_sequences else 2  # type: ignore[assignment]
+        yaml.sequence_dash_offset = yaml.sequence_indent - 2  # type: ignore[operator]
+
+        if self.yaml_indent_sequences:  # in the future: or other formatting options
+            # For root-level sequences, FormattedEmitter overrides sequence_indent
+            # and sequence_dash_offset to prevent root-level indents.
+            yaml.Emitter = FormattedEmitter
+
+        self._yaml = yaml
+        return yaml
+
+    def run(self) -> None:
+        """For each file, read it, execute fmt transforms on it, then write it."""
+        for file, matches in self.matches_per_file.items():
+            # str() convinces mypy that "text/yaml" is a valid Literal.
+            # Otherwise, it thinks base_kind is one of playbook, meta, tasks, ...
+            file_is_yaml = str(file.base_kind) == "text/yaml"
+
+            try:
+                data: Union[CommentedMap, CommentedSeq, str] = file.content
+            except (UnicodeDecodeError, IsADirectoryError):
+                # we hit a binary file (eg a jar or tar.gz) or a directory
+                data = ""
+                file_is_yaml = False
+
+            if file_is_yaml:
+                # ruamel.yaml only preserves empty (no whitespace) blank lines.
+                # So, drop spaces in whitespace-only lines. ("\n  \n" -> "\n\n")
+                data = _whitespace_only_lines_re.sub("", cast(str, data))
+                # load_data has an lru_cache, so using it should be cached vs using YAML().load() to reload
+                data = load_data(data)
+                self.yaml.dump(data, file.path)
```

Comment:
Note that `yaml.dump()` has optimizations around creating the string & dumping to file all at once. The docs really push people away from separating creating a string and then writing it to a file:
https://yaml.readthedocs.io/en/latest/example.html#output-of-dump-as-a-string

There is no `dumps()` function w/ `ruamel.yaml`. Though a `dumps()` function might make tests that avoid IO easier, it's going to be a bunch of work to get around the design decisions of `ruamel.yaml` to write to a `StringIO()` instead of directly to the file. If there's a clean way to do this, I'm all ears.

Reply:
I think it's good. Sure, it's a bit less nice than `dumps`, but it's only a few extra lines in a test.
I think the way you've used the StringIO in the tests provides a good seam for testing. In my head I was thinking that we should have a function to handle the formatting and dump to File or StringIO, and then I wrote it out as
```python
class Transformer:
    ...
    def dumps(self, data, dest)->None:
        self.yaml.dump(data, dest)
```
which really isn't adding anything of value.

So I think the way you've done it here is good.